addons/udes_routing/tests/test_push_from_drop.py

```diff
@@ -0,0 +1,240 @@
+                         {'product': self.banana, 'uom_qty': 99}]
+        self.goods_in = self.create_picking(self.picking_type_goods_in,
+                                            products_info=products_info, confirm=True, assign=True)
+
+        self.move_lines = self.goods_in.move_line_ids
+        for ml in self.move_lines:
+            ml.write({'qty_done': ml.product_uom_qty})
+
+    def test_picking_has_correct_src_and_dest(self):
+        """Test that pickings created by push from drop have the src and dest
+        location of the rule that created them."""
+        for ml in self.move_lines:
+            ml.write({'location_dest_id': self.received_damaged_location.id})
+        putaway = self._complete_picking_and_return_next_picking_id()
+        self.assertEqual(len(putaway), 1)
+        self.assertEqual(putaway.location_id, self.received_damaged_location)
+        self.assertEqual(putaway.location_dest_id, self.location_qc_zone)
+
+    def test_no_push_rule(self):
+        """Test that when no push rule exists nothing is created"""
+        (self.push_damaged_putaway + self.push_putaway).unlink()
+        putaway = self._complete_picking_and_return_next_picking_id()
+        self.assertEqual(len(putaway), 0)
+
+    def test_stock_is_reserved(self):
+        """Test that when a push occurs, stock is reserved and available to move
+        further."""
+        putaway = self._complete_picking_and_return_next_picking_id()
+        self.assertEqual(len(putaway), 1)
+        self.assertEqual(putaway.state, 'assigned')
+
+    def test_picking_can_be_completed(self):
+        """Test that when a picking is created by push_from_drop it can be
+        validated."""
+        putaway = self._complete_picking_and_return_next_picking_id()
+        self.assertEqual(len(putaway), 1)
+        self.assertEqual(putaway.state, 'assigned') 
+        for ml in putaway.move_line_ids:
+            ml.write({'qty_done': ml.product_uom_qty,
+                      'location_dest_id': self.location_qc_01.id})
+        putaway._action_done()
+        self.assertEqual(putaway.state, 'done')
+
+    def test_picking_info_propagated(self):
+        """Test that when a picking is created by push_from_drop it contains
+        the origin and partner of the source picking."""
+        new_partner = self.create_partner('Test partner')
+        self.goods_in.partner_id = new_partner
+        new_origin = 'Test origin'
+        self.goods_in.origin = new_origin
+        putaway = self._complete_picking_and_return_next_picking_id()
+        self.assertEqual(putaway.origin, new_origin)
+        self.assertEqual(putaway.partner_id, new_partner)
+
+    def _complete_picking_and_return_next_picking_id(self):
+        self._set_picking_destination_location()
+        self.goods_in._action_done()
+        return self.goods_in.u_next_picking_ids
+    
+    def _set_picking_destination_location(self):
+        for ml in self.move_lines:
```

Comment:
nit: Can place these writes in _complete_picking_and_return_next_picking_id as looks like only used in there